Centralize deployment targets into `xcconfig` files

mokagio opened the pull request:
<!-- Remember about a good descriptive title. -->


<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
In https://github.com/woocommerce/woocommerce-ios/pull/15000#discussion_r1942595988, @AliSoftware noticed I removed a leftover definition of the iOS deployment target being 15.0, when it had already been discussed, approved, and implemented (#11867) that we should target 16.0.

@AliSoftware recommended running it past the team — @woocommerce/apple-developers? @woocommerce/ios-developers? —  that the deployment target for the whole `WooCommerce.xcodeproj` was now 16.0.

This PR does that and more, by DRYing the definition for the iOS, macOS, and watchOS deployment target into a single `xcconfig`, `Common.xcconfig`.

**Before:**

```
$ git grep 'TARGET ='

Experiments/Experiments.xcodeproj/project.pbxproj:				IPHONEOS_DEPLOYMENT_TARGET = 16.0;
Experiments/Experiments.xcodeproj/project.pbxproj:				IPHONEOS_DEPLOYMENT_TARGET = 16.0;
Experiments/Experiments.xcodeproj/project.pbxproj:				IPHONEOS_DEPLOYMENT_TARGET = 16.0;
Fakes/Fakes.xcodeproj/project.pbxproj:				IPHONEOS_DEPLOYMENT_TARGET = 16.0;
Fakes/Fakes.xcodeproj/project.pbxproj:				IPHONEOS_DEPLOYMENT_TARGET = 16.0;
Hardware/Hardware.xcodeproj/project.pbxproj:				IPHONEOS_DEPLOYMENT_TARGET = 16.0;
Hardware/Hardware.xcodeproj/project.pbxproj:				IPHONEOS_DEPLOYMENT_TARGET = 16.0;
Hardware/Hardware.xcodeproj/project.pbxproj:				IPHONEOS_DEPLOYMENT_TARGET = 16.0;
Networking/Networking.xcodeproj/project.pbxproj:				IPHONEOS_DEPLOYMENT_TARGET = 16.0;
Networking/Networking.xcodeproj/project.pbxproj:				MACOSX_DEPLOYMENT_TARGET = 14.0;
Networking/Networking.xcodeproj/project.pbxproj:				WATCHOS_DEPLOYMENT_TARGET = 9.0;
Networking/Networking.xcodeproj/project.pbxproj:				MACOSX_DEPLOYMENT_TARGET = 14.0;
Networking/Networking.xcodeproj/project.pbxproj:				WATCHOS_DEPLOYMENT_TARGET = 9.0;
Networking/Networking.xcodeproj/project.pbxproj:				MACOSX_DEPLOYMENT_TARGET = 14.0;
Networking/Networking.xcodeproj/project.pbxproj:				WATCHOS_DEPLOYMENT_TARGET = 9.0;
Networking/Networking.xcodeproj/project.pbxproj:				IPHONEOS_DEPLOYMENT_TARGET = 16.0;
Networking/Networking.xcodeproj/project.pbxproj:				IPHONEOS_DEPLOYMENT_TARGET = 16.0;
Storage/Storage.xcodeproj/project.pbxproj:				IPHONEOS_DEPLOYMENT_TARGET = 16.0;
Storage/Storage.xcodeproj/project.pbxproj:				IPHONEOS_DEPLOYMENT_TARGET = 16.0;
Storage/Storage.xcodeproj/project.pbxproj:				IPHONEOS_DEPLOYMENT_TARGET = 16.0;
WooCommerce/WooCommerce.xcodeproj/project.pbxproj:				WATCHOS_DEPLOYMENT_TARGET = 9.0;
WooCommerce/WooCommerce.xcodeproj/project.pbxproj:				WATCHOS_DEPLOYMENT_TARGET = 9.0;
WooCommerce/WooCommerce.xcodeproj/project.pbxproj:				WATCHOS_DEPLOYMENT_TARGET = 9.0;
WooCommerce/WooCommerce.xcodeproj/project.pbxproj:				WATCHOS_DEPLOYMENT_TARGET = 9.0;
WooCommerce/WooCommerce.xcodeproj/project.pbxproj:				WATCHOS_DEPLOYMENT_TARGET = 9.0;
WooCommerce/WooCommerce.xcodeproj/project.pbxproj:				WATCHOS_DEPLOYMENT_TARGET = 9.0;
WooCommerce/WooCommerce.xcodeproj/project.pbxproj:				WATCHOS_DEPLOYMENT_TARGET = 9.0;
WooCommerce/WooCommerce.xcodeproj/project.pbxproj:				WATCHOS_DEPLOYMENT_TARGET = 9.0;
WooCommerce/WooCommerce.xcodeproj/project.pbxproj:				WATCHOS_DEPLOYMENT_TARGET = 9.0;
WooFoundation/WooFoundation.xcodeproj/project.pbxproj:				WATCHOS_DEPLOYMENT_TARGET = 9.0;
WooFoundation/WooFoundation.xcodeproj/project.pbxproj:				WATCHOS_DEPLOYMENT_TARGET = 9.0;
WooFoundation/WooFoundation.xcodeproj/project.pbxproj:				WATCHOS_DEPLOYMENT_TARGET = 9.0;
WooFoundation/WooFoundation.xcodeproj/project.pbxproj:				IPHONEOS_DEPLOYMENT_TARGET = 16.0;
WooFoundation/WooFoundation.xcodeproj/project.pbxproj:				IPHONEOS_DEPLOYMENT_TARGET = 16.0;
WooFoundation/WooFoundation.xcodeproj/project.pbxproj:				IPHONEOS_DEPLOYMENT_TARGET = 16.0;
Yosemite/Yosemite.xcodeproj/project.pbxproj:				IPHONEOS_DEPLOYMENT_TARGET = 16.0;
Yosemite/Yosemite.xcodeproj/project.pbxproj:				IPHONEOS_DEPLOYMENT_TARGET = 16.0;
Yosemite/Yosemite.xcodeproj/project.pbxproj:				IPHONEOS_DEPLOYMENT_TARGET = 16.0;
config/Common.xcconfig:IPHONEOS_DEPLOYMENT_TARGET = 16.0
```

**After:**

```
$ git grep 'TARGET ='

config/Common.xcconfig:IPHONEOS_DEPLOYMENT_TARGET = 16.0
config/Common.xcconfig:MACOSX_DEPLOYMENT_TARGET = 14.0
config/Common.xcconfig:WATCHOS_DEPLOYMENT_TARGET = 9.0
```

<!-- blue -->
> [!NOTE]  
> As a byproduct of moving the deployment target setting that _some_ of the targets had into a common file, now _all_ targets have a specific watchOS and macOS deployment target. As far as I understand, that's not a problem. As a matter of fact, it should ensure consistency in the future.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
N.A.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->
If CI is green, then the change did not affect the build process.

Also notice that if you now change one or more of the values in `Common.xcconfig` the changes will reflect across all targets.

![image](https://github.com/user-attachments/assets/fad51210-2757-457a-b10b-c396255b28c3)

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

mokagio committed:
Centralize WooCommerce `WATCHOS_DEPLOYMENT_TARGET` setting to `xcconfig`

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

mokagio committed:
Set up `xcconfig` for Yosemite and push iOS version there

mokagio committed:
Set up `xcconfig` for Storage and push iOS version there

mokagio committed:
Set up `xcconfig` for Networking and push iOS version there

mokagio committed:
Set up `xcconfig` for Experiments and push iOS version there

mokagio committed:
Set up `xcconfig` for Hardware and push iOS version there

mokagio committed:
Set up `xcconfig` for WooFoundation and push iOS version there

mokagio committed:
Set up `xcconfig` for Fakes and push iOS version there

mokagio committed:
Rename `Common.xcconfig` to `WooCommerce.common.xcconfig`

mokagio committed:
DRY macOS and watchOS deployment targets for Networking

mokagio committed:
DRY watchOS deployment target for WooFoundation

mokagio committed:
DRY iOS, macOS, and watchOS deployment targets to new `Common.xcconfig`

mokagio committed:
Add note on how `Common.xcconfig` works

AliSoftware commented on Experiments/Experiments.xcodeproj/project.pbxproj:
```diff
@@ -247,6 +250,7 @@
 			isa = PBXResourcesBuildPhase;
 			buildActionMask = 2147483647;
 			files = (
+				3FA7D9FC2D547E1700CE5611 /* Experiments.xcconfig in Resources */,
```
The `.xcconfig` files should not appear as `PBXBuildFile` in the project nor be added to `PBXResourcesBuildPhase`: there's no point in including the `.xcconfig` file as a resource inside the final `.app` bundle, as it's only a config file used during the build.

(Of course, same remark applies to all the `.xcconfig` files you added)

mokagio replied:
Of course. Thank you for noticing this, since I had obviously missed it.

I fixed it in 655c41c.

The thing that surprises me is that Xcode added them only to _three_ of the target/projects I touched. What's the logic behind that? I'm pretty sure I used the same approach to add the file to each: select the framework target, cmd n, add the file 🤔 🤷‍♂️

mokagio committed:
Remove `xcconfig`s from the copy bundle resources build phase

Xcode put them there, but it did it only for some of the `xcconfig`s.
Why? What was the logic it used?

AliSoftware reviewed (approved):
Approving that the changes in the Xcode setup look good.

Though you should probably still wait for an approval from someone from @woocommerce/ios-developers to confirm the consistent bump to `IPHONEOS_DEPLOYMENT_TARGET = 16.0` is ok by them too before merging still 😉

mokagio commented:
Thanks @AliSoftware !

Yep, happy to wait for okay from the team. I shared this internally, too, p1738873977098429-slack-C6H8C3G23

mokagio commented:
The Slack ping to the dev team got three 👍 and one positive reply. Merging.
